doc: gatsby image configuration (#652)
johnsonandjohnson/Bodiless-JS: packages/bodiless-documentation/doc/Development/Guides/SiteBuildBasics.md

)();

> ```

Reviewer: @hvanyo i agree that 'Building Sites' might be a better title. Or maybe 'Building a Basic Site'? Will ask Kevin and Sophia.

> ```

Reviewer 2: @malachye  @beliayeu  just to confirm -- this discussion should go into different ticket and is not relevant to this PR.

## Image Configurations

BodilessJS provides integration with [gatsby-image](https://www.gatsbyjs.com/plugins/gatsby-image/) plugin for creating optimized images and leveraging advanced image loading techniques.

### Activation

1. Create a HOF on site level responsible for transforming a plain image based component into gatsby image component.

```ts
import {
  asGatsbyImage as asBaseGatsbyImage,
  withGatsbyImageNode,
  withGatsbyImageLogger,
} from '@bodiless/gatsby-theme-bodiless';
import { asBodilessImage } from '@bodiless/components-ui';

const asGatsbyImg = (preset: string) => flowRight(
  withGatsbyImageNode(preset),
  asBodilessImage(),
  withGatsbyImageLogger(preset),
  asBaseGatsbyImage,
);
```

> > Note: We are using `asGatsbyImage`, `withGatsbyImageNode`, `withGatsbyImageLogger` functions provided by `@bodiless/gatsby-theme-bodiless`.

Reviewer: ```suggestion
> Note: BodilessJS uses `asGatsbyImage`, `withGatsbyImageNode`, `withGatsbyImageLogger` functions provided by `@bodiless/gatsby-theme-bodiless`.
```

> * `asGatsbyImage` is a HOC that either replaces the component with `GatsbyImg` in case of data required for `GatsbyImg` is available or it renders the input component otherwise.
> * `withGatsbyImageNode` is a HOF that adds a gatsby image BodilessJS node, which enriches image node data with image preset provided as an input.
> * `withGatsbyImageLogger` is a HOF that fails gatsby build and logs errors when there is a mismatch between image preset passed as an argument to gatsby image node and the corresponding image preset stored in the image node json file.

1. Create a HOC for each image preset, for instance:

```ts
import { GatsbyImagePresets } from '@bodiless/gatsby-theme-bodiless';

const asFluidGatsbyImage = asGatsbyImg(GatsbyImagePresets.Fluid);
const asFluidWithWebpGatsbyImage = asGatsbyImg(GatsbyImagePresets.FluidWithWebp);
);
```

> > Note: We are using default image presets exposed by `@bodiless/gatsby-theme-bodiless`. Visit [default image presets](#default-image-presets) section to get a list of available presets.

Reviewer: ```suggestion
> Note: BodilessJS uses default image presets exposed by `@bodiless/gatsby-theme-bodiless`. Visit [default image presets](#default-image-presets) section to get a list of available presets.
```

1. Use the HOCs to compose gatsby image components

```ts
import { Img } from '@bodiless/fclasses';

const FluidGatsbyImage = asFluidGatsbyImage(Img);
const FluidWithWebpGatsbyImage = asFluidWithWebpGatsbyImage(Img);
```

1. Render the gatsby image components as any other React component.

```tsx
 <FluidGatsbyImage />
```

1. Upload images via BodilessJS admin interface.

> Note: when you activate gatsby image for an existing image, then you need to reupload the image or update the corresponding image json file.

### Change image preset

> Lest consider, you have an image with `GatsbyImagePresets.Fluid` preset and you want to change the preset to `GatsbyImagePresets.FluidNoBase64`.

Reviewer: ```suggestion
If you have an image with `GatsbyImagePresets.Fluid` preset and you want to change the preset to `GatsbyImagePresets.FluidNoBase64`.
```
